src/Compilers/CSharp/Portable/Syntax/NonRecursiveSyntaxRewriter.cs

```diff
@@ -0,0 +1,286 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Represents a non-recursive visitor which descends an entire <see cref="CSharpSyntaxNode"/> graph and
+    /// may replace or remove visited SyntaxNodes in depth-first order.
+    /// </summary>
+    public partial class NonRecursiveSyntaxRewriter : CSharpSyntaxVisitor<NonRecursiveSyntaxRewriter.Chunk>
```

Comment:
No because with my rewriter, we have 2 methods: Visit and Transform. 
In Visit, you can short-circuit the Transform or one part of the children transformation.

However, that's right that you cannot do Visit in many steps as we can with CSharpSyntaxRewriter.
e.g.:
```
var left = Visit(binary.Left);
if (<condition based on left>)
{
   Visit(binary.right)
}
```